src/coreclr/jit/importercalls.cpp

```diff
@@ -3786,6 +3789,32 @@ GenTree* Compiler::impIntrinsic(CORINFO_CLASS_HANDLE    clsHnd,
                 break;
             }
 
+            case NI_System_ArgumentNullException_ThrowIfNull:
+            {
+                // void ThrowIfNull(object? argument, string? paramName = null)
+                // void ThrowIfNull(object? argument, ExceptionArgument paramName)
+                assert(sig->numArgs == 2);
+                assert(sig->retType == CORINFO_TYPE_VOID);
+
+                // if we see:
+                //
+                //   ArgumentNullException_ThrowIfNull(GT_BOX(...), ...)
```

Comment:
Will this work correctly for `Nullable<>`?

Reply:
> Will this work correctly for `Nullable<>`?

Ah, good question - it does work correctly now, but doesn't remove the boxing for e.g. `Test<int?>`, let me think

Reply:
hm.. I don't see how we can easily handle Nullable<> in Tier0 unfortunately 🙁 

for non-nullable we do something like this:
```
ArgumentNullException.ThrowIfNull(BOX(val), name);
```
is transformed into:
```
_ = val; // preserve side-effects
_ = name;  // preserve side-effects
```
For Nullable it won't work, we have this:
```
ArgumentNullException.ThrowIfNull(CORINFO_HELP_BOX_NULLABLE(cls, val), name);
```
and in this case we can't remove the call - it is still needed in case if our `Nullable<T>._hasValue` is `false`. So in theory we can import it like this:
```
var valTmp = val;
var nameTmp = name;
if (!valTmp._hasValue)
    ArgumentNullException.ThrowIfNull(null, nametmp);
```
but I don't see a simple way to perform such a transformation in Tier0, it's definitely not in importer (and not in morph).
And no, we can't emit our own `CORINFO_HELP_THROW_NRE` becasue we still need to preserve `ThrowIfNull`'s behavior (e.g. it passes arg name (or ExceptionValue)).

any ideas @AndyAyersMS @jakobbotsch @dotnet/jit-contrib ?

Reply:
You could create a `QMARK` for it?

Reply:
> You could create a `QMARK` for it?

Not in `impIntrinsic` since the actual method is not yet materialized, but maybe I can do it when it is 🤔

Reply:
QMARK, or defer until you can create flow.

Reply:
(If we can't address nullable, it's not a big deal.)

Reply:
> (If we can't address nullable, it's not a big deal.)

I've already addressed 🙂